### DISCUSSION TITLE
RFC: Improve Platform CI in edk2

### DISCUSSION THREAD

**Discussion by os-d** [tianocore/edk2#6364] [General]
There is an ongoing effort to improve edk2's CI processes, in particular the length of time spent on PR gates. There are currently ~90 CI jobs that run on every PR and about half of them are from three packages: ArmVirtPkg, EmulatorPkg, and OvmfPkg. The reason these packages contribute so many jobs to CI is that there are many platforms within each package that each run CI in every PR.

The request from the edk2 CI working group is to get feedback from the platform maintainers on how we can reduce the overall number of jobs from these three packages while still running valuable checks and keeping overall run time to a minimum.

The weekly tools and CI meeting has been discussing this topic, which some of the platform maintainers have been part of. We would like to get feedback from the maintainers by the next instance [Oct. 28th at 8am PST](https://edk2.groups.io/g/devel/viewevent?eventid=2509145), maintainers are also welcome to join this meeting to discuss. Please use this discussion to record feedback prior to then.

The CI working group will take this feedback and make a set of proposals. In conjunction with this, the group is looking at reducing and optimizing other jobs.

At the [10/21 meeting](https://github.com/tianocore/edk2/discussions/6356) the working group came up with a list of questions to solicit feedback on:

- What are the most essential builds necessary to include in PR status checks?
- What are the most necessary checks to run for the overall package? Can checks happening in different steps/jobs today be combined in the future?
- Do individual platforms (e.g. bhyve, Xen, etc.) under ArmVirt/OVMF need to run on PR gates or could they be nightly CI?
- Would some of these individual platforms be better homed in edk2-platforms?
- Can you provide CI agents to reduce the impact of your platform CI on the overall project's agents?

To give more clarity and context to this list, the CI working group has been discussing new methods of CI in edk2. We may choose to move some current PR status checks to a nightly CI run to reduce load on PR gates. Another proposal was reducing the set of platforms in edk2 to a more minimal, core set (with others migrating to edk2-platforms). This would ensure that edk2 can boot in what is considered a core virtual environment but does not check each flavor under ArmVirt and OVMF, as by and large these are similar and if one fails, others are likely to.

Again, there are other optimizations being discussed, this discussion is just focused on the platform side as that contributes half of the jobs running in each PR gate and we want to ensure each platform maintainer has a chance to weigh in.

**Comment by os-d** [comment]
@'ing platform maintainers/reviewers here to draw attention:

ArmVirtPkg: @samimujawar @leiflindholm @kraxel 

EmulatorPkg: @ajfish @niruiyu 

OvmfPkg: @jyao1 @kraxel 

bhyve: @bexcran @corvink

cloudhv: @jongwu @weltling 

microvm: @kraxel 

AmdSev: @ruleof2 @jyao1 @mxu9 @tlendacky @mdroth 

Xen: @tperard 

RiscVVirt: @vlsunil @andreiw 

LoongArchQemu: @kilaterlee @bibo-mao @lixianglai

**Comment by bexcran** [comment]
Having maintained a company's CI servers for a few years and worked to reduce the time for CI to run from 20 minutes to 5, I like to think I have some experience around CI.

Personally, I think much of what we're doing at the moment is wrong. It feels like the standard GitHub runners are _slow_, while we have "Larger GitHub-hosted runners" unprovisioned. We appear to be working around that by _also_ using Azure, but that's pretty confusing for anyone looking at the CI system.

Other significant open source projects such as FreeBSD and coreboot use Jenkins and don't require everything to be in "the cloud" (i.e. somebody else's computer). I've already offered use of my 128-core AMD EPYC server, and I also have a 96-core Ampere Altra machine I could put in a colo. 

In terms of pre-commit vs post-commit vs nightly, I think:

Pre-commit should check for style and spelling issues, which should be very quick. Then do a set of build tests - how many I'm not sure, but probably not multiple per top-level package?  Post-commit, we can run a full firmware build including ideally posting firmware images for OVMF, ArmPkg etc. Then nightly we can run more time-consuming checks such as Coverity, CodeQL etc. We might also have tasks that should be done weekly, such as making sure every single combination of toolchain/OS/architecture builds.

We should also consider how edk2-platforms fits into this, since I don't think it should be a second-class citizen.

**Comment by olegilyasov** [comment]
![Screenshot 2024-10-17 075626](https://github.com/user-attachments/assets/c1d79601-bf1c-47fa-a0dd-bfc75c4753d4)
Note: when a developer hits "Update branch" button, github creates a commit with a message that will not pass the CI (sign-off will be missing).

**Reply by mdkinney** [reply]
There is an option to Rebase in the drop down.  Always select that.  PatchCheck fails if there is a merge commit.

**Reply by bexcran** [reply]
Do these settings also affect updating a branch? Or is it not possible to disallow merge commit when updating?
![Uploading Screenshot_2024-10-23_09-25-45.png…]()

**Reply by mdkinney** [reply]
I am not able to find a setting to disallow merge in that button in the web UI.  The github settings for a repo can require linear history, so that prevents merge commits from being added to repo branches.  That setting, however, does not prevent merge commits in a PR.

**Reply by makubacki** [reply]
@mdkinney, if this causes more confusion than it helps, you could simply turn it off and force people to rebase locally.

Other projects I'm involved in allow merge types other than rebase (still linear - such as squash) where this is helpful. It could arguably be unhelpful here if it's putting developers in a position where they unknowingly fail CI checks by using it.

**Comment by os-d** [comment]
@'ing platform maintainers/reviewers here to draw attention again:

ArmVirtPkg: @samimujawar @leiflindholm @kraxel

EmulatorPkg: @ajfish @niruiyu

OvmfPkg: @jyao1 @kraxel

bhyve: @bexcran @CorvinK

cloudhv: @jongwu @weltling

microvm: @kraxel

AmdSev: @ruleof2 @jyao1 @mxu9 @tlendacky @mdroth

Xen: @tperard

RiscVVirt: @vlsunil @andreiw

LoongArchQemu: @kilaterlee @bibo-mao @lixianglai

Hi platform maintainers,

We have not gotten feedback from platform maintainers except for Rebecca. We would like to get any feedback. Our next steps are to make a proposed set of changes to PR gate CI re:platform CI builds in PR form and have the platform maintainers/reviewers added to the PR. If there is no feedback from maintainers on the PR, we will check it in and will await further feedback from maintainers. I am going to try a few other channels to reach platform maintainers.

If you don't have any opinion on the subject, please let us know in this forum as well so that we can be aware you saw this and have no opinion.

**Reply by kilaterlee** [reply]
Sorry for the late response.
Currently, the CI checks what seems to be quite a lot of projects on Github and Azure, so it's pretty slow to check every PR, but I think the items in Github and Azure are necessary.

About the night CI. EDK2 is a global project, I think it's not adapt everyone, because the time are different in each continents. :)

This is my idea, but I think it is not mature yet: If CI more smarter, we can separate the public module and platform module checks in CI: 
If MdePkg or UefiCpuPkg has changed (public module), only the build, ECC and other checks of this module(or associated modules) will be run, and other public modules do not need CI checks. And we choose a typical platform to check whether the changed public module can run correctly.
If the platform module has changed(ArmPkg/ArmVirtPkg/OvmfPkg etc.), only check this module, like build, ECC, execute etc., and public modules do not need CI checks.

I think it can improve CI efficiently. We should run a full CI check, maybe once a week or half a month, and must ensure the full CI check passes before edk2 is released.

About the edk2-platform, I think I could accept homed the individual platforms into it, but only if edk2-platform had the same CI checks as the edk2 repo. Just like Rebecca saied: edk2-platform shouldn't be a second-class citizen.

And I have a question bellow:


> @'ing platform maintainers/reviewers here to draw attention again:
> 
> ArmVirtPkg: @samimujawar @leiflindholm @kraxel
> 
> EmulatorPkg: @ajfish @niruiyu
> 
> OvmfPkg: @jyao1 @kraxel
> 
> bhyve: @bexcran @CorvinK
> 
> cloudhv: @jongwu @weltling
> 
> microvm: @kraxel
> 
> AmdSev: @ruleof2 @jyao1 @mxu9 @tlendacky @mdroth
> 
> Xen: @tperard
> 
> RiscVVirt: @vlsunil @andreiw
> 
> LoongArchQemu: @kilaterlee @bibo-mao @lixianglai
> 
> Hi platform maintainers,
> 
> We have not gotten feedback from platform maintainers except for Rebecca. We would like to get any feedback. Our next steps are to make a proposed set of changes to PR gate CI re:platform CI builds in PR form and have the platform maintainers/reviewers added to the PR. If there is no feedback from maintainers on the PR, we will check it in and will await further feedback from maintainers. I am going to try a few other channels to reach platform maintainers.

IIUC. Do the platform maintainer need to commit a PR to trigger platform CI checks?
> 
> If you don't have any opinion on the subject, please let us know in this forum as well so that we can be aware you saw this and have no opinion.

**Comment by samimujawar** [comment]
I think for ArmVirtPkg we can have nightly builds. I understand that timezone differences can be an issue. However, having regular nightly builds will help catch issues earlier.

**Reply by leiflindholm** [reply]
ArmVirtPkg, like OvmfPkg is the only way we get complete images built.

Like for OvmfPkg, we could look at reducing the number of builds that are PR gating vs those that can be deferred to nightly, but we need some ArmVirtPkg builds to remain gating.

**Comment by os-d** [comment]
Thanks @kilaterlee and @samimujawar for weighing in. Definitely agreed that the term nightly here is not quite appropriate, we could also call it daily :). The central idea being that not all of these checks may need to be run on every PR and instead can be run periodically to catch issues.

We will need to define a process on what happens when a daily/nightly (or other) build fails. The goal would be to have one representative of ArmVirt/Ovmf be run on PRs, which should catch the general cases and then the subflavors only run periodically, since they typically align with the main ArmVirt or Ovmf build.

@kilaterlee , the platform maintainer does not need to commit a PR to trigger platform CI checks. My statement there was saying that if we do not hear from platform maintainers, the CI subgroup will put up a PR redefining the PR gates as we see fit and ask all platform maintainers for review. If they do not review that, we will assume they do not wish to weigh in on the subject and move forward with our plan.

**Reply by bexcran** [reply]
> We will need to define a process on what happens when a daily/nightly (or other) build fails.

Ideally we'd send an email/message/ping to anyone who'd committed code since the last successful build.

**Comment by makubacki** [comment]
I want to refocus this particular "platform" thread. Let me clarify a few things about this discussion versus the larger CI discussion.

---

- We have ~90 CI jobs scheduled for CI run today - this included "PR gates" that run on PR updates and "Post-Merge CI" that run after a PR is merged.
- Half of those come from the platforms outlined at the beginning of this discussion.
- There are other improvements to CI being discussed see the Tianocore Tools & CI meeting notes in the Discussions area and the ideas here https://github.com/tianocore/edk2/discussions/6017. Feel free to add any generic ideas you have there.

We have a total of 30 parallel agents.

## The unique perspective needed from reaching out to platform owners here is:

1. **What do you need to run for your platform?**
2. **When do you need to run it?**

We are balancing "core" versus "package" utilization of CI resources and we need this information to determine how much to allocate to platforms. Ideally, we need all platforms to schedule a combined maximum of 10 jobs (out of the 30) so we do not oversaturate build agents.

Pushing the final allocation of resources, generic improvements to CI, and other topics aside. For your platform, can you please list what checks you need to run on every PR update (for code that affects your platform) and what checks you think can be moved to a less regular schedule such as "Post-Merge" (not run in a PR at all), "Nightly" (aka daily), "Other" (custom that you mention)?

**Reply by bexcran** [reply]
For me (Bhyve), I don't need any CI run.

**Comment by kraxel** [comment]
Well, one reason for OvmfPkg and ArmVirtPkg being built is that it these are complete firmware images, so you'll get some end-to-end testing, and with the OVMF images being booted to efi shell even a basic smoke test to see whenever the image actually works.  I think this is a check worth keeping in PR CI runs, because changes outside OvmfPkg can break OVMF, and in case that happens there is a high chance it breaks other platforms too.

There is a large number of OVMF configurations though, and I think we do not need them all to archive that effect.  For example the xen / microvm / bhyve build configurations could be separated.  Run them only on PRs which actually change OvmfPkg code.

**Reply by makubacki** [reply]
Thanks @kraxel. You're right that we'd like to keep those basic boot tests in PRs.

@bexcran, has already responded that bhyve does not need to run. To help us get to an actionable change in the pipelines, can you or others please help narrow the list of existing jobs to what you'd like to see run in pull requests?

Today the following pipelines run. We could either identify jobs to eliminate or keep. Since we are trying to greatly reduce the total number of platform jobs run in pull requests, could you please provide a list of jobs you think are essential to keep there? Others feel free to reply as well.

- Pipeline: PlatformCI_OvmfPkg_Windows_VS2019_PR
  - https://dev.azure.com/tianocore/edk2-ci/_build?definitionId=51
  - Number of jobs spawned: 11
  - Jobs:
    - OVMF IA32 DEBUG (~16 min)
    - OVMF IA32 RELEASE (~16 min)
    - OVMF IA32 NOOPT (~13 min)
    - OVMF X64 DEBUG (~14 min)
    - OVMF X64 RELEASE (~15 min)
    - OVMF X64 NOOPT (~20 min)
    - OVMF IA32X64 DEBUG (~14 min)
    - OVMF IA32X64 RELEASE (~12 min)
    - OVMF IA32X64 NOOPT (~12 min)
    - OVMF IA32X64 FULL DEBUG (~23 min)
    - OVMF IA32X64 FULL RELEASE (~18 min)
- PlatformCI_OvmfPkg_Ubuntu_GCC5_PR
  - https://dev.azure.com/tianocore/edk2-ci/_build?definitionId=49
  - Number of jobs spawned: 19
  - Jobs:
    - OVMF IA32 DEBUG (~6 min)
    - OVMF IA32 RELEASE (~6 min)
    - OVMF IA32 NOOPT (~6 min)
    - OVMF X64 DEBUG (~9 min)
    - OVMF X64 RELEASE (~9 min)
    - OVMF X64 NOOPT (~6 min)
    - OVMF IA32X64 DEBUG (~8 min)
    - OVMF IA32X64 RELEASE (~7 min)
    - OVMF IA32X64 NOOPT (~6 min)
    - OVMF IA32X64 FULL DEBUG (~12 min)
    - OVMF IA32X64 FULL RELEASE (~11 min)
    - AMDSEV X64 DEBUG (~7 min)
    - BHYVE X64 DEBUG (~8 min)
    - CLOUDHV X64 DEBUG (~8 min)
    - MICROVM X64 DEBUG (~7 min)
    - OVMF XEN X64 DEBUG (~9 min)
    - INTELTDX X64 DEBUG (~8 min)
    - RISVC64 DEBUG (~6 min)
    - LOONGARCH64 DEBUG (~7 min)
- PlatformCI_EmulatorPkg_Windows_VS2019_PR
  - https://dev.azure.com/tianocore/edk2-ci/_build?definitionId=45
  - Number of jobs spawned: 12
  - Jobs:
    - EmulatorPkg X64 DEBUG (~10 min)
    - EmulatorPkg X64 RELEASE (~11 min)
    - EmulatorPkg X64 NOOPT (~10 min)
    - EmulatorPkg IA32 DEBUG (~12 min)
    - EmulatorPkg IA32 RELEASE (~11 min)
    - EmulatorPkg IA32 NOOPT (~12 min)
    - EmulatorPkg X64 FULL DEBUG (~15 min)
    - EmulatorPkg X64 FULL RELEASE (~11 min)
    - EmulatorPkg X64 FULL NOOPT (~13 min)
    - EmulatorPkg IA32 FULL DEBUG (~12 min)
    - EmulatorPkg IA32 FULL RELEASE (~15 min)
    - EmulatorPkg IA32 FULL NOOPT (~11 min)
- PlatformCI_EmulatorPkg_Ubuntu_GCC5_PR
  - https://dev.azure.com/tianocore/edk2-ci/_build?definitionId=42
  - Number of jobs spawned: 6
  - Jobs:
    - EmulatorPkg X64 DEBUG (~7 min)
    - EmulatorPkg X64 RELEASE (~6 min)
    - EmulatorPkg X64 NOOPT (~5 min)
    - EmulatorPkg X64 FULL DEBUG (~7 min)
    - EmulatorPkg X64 FULL RELEASE (~8 min)
    - EmulatorPkg X64 FULL NOOPT (~5 min)
- PlatformCI_ArmVirtPkg_Ubuntu_GCC5_PR
  - https://dev.azure.com/tianocore/edk2-ci/_build?definitionId=47
  - Number of jobs spawned: 18
  - Jobs:
    - QEMU AARCH64 DEBUG (~9 min)
    - QEMU AARCH64 RELEASE (~8 min)
    - QEMU AARCH64 NOOPT (~6 min)
    - QEMU ARM DEBUG (~7 min)
    - QEMU ARM RELEASE (~8 min)
    - QEMU ARM NOOPT (~6 min)
    - QEMU KERNEL AARCH64 DEBUG (~7 min)
    - QEMU KERNEL AARCH64 RELEASE (~6 min)
    - QEMU KERNEL AARCH64 NOOPT (~7 min)
    - QEMU KERNEL ARM DEBUG (~6 min)
    - QEMU KERNEL ARM RELEASE (~8 min)
    - QEMU KERNEL ARM NOOPT (~7 min)
    - KVMTOOL AARCH64 DEBUG (~6 min)
    - KVMTOOL AARCH64 RELEASE (~6 min)
    - KVMTOOL ARM DEBUG (~6 min)
    - KVMTOOL ARM RELEASE (~6 min)
    - CLOUDHV AARCH64 DEBUG (6 min)
    - CLOUDHV AARCH64 RELEASE (~7 min)

**Reply by mdkinney** [reply]
Can the typical execution time for these jobs be added?  I think EmulatorPkg builds are faster than OcmfPkg build.  I also think GCC builds are typically faster than VS20xx builds.  For proposals to reduce builds, it would be good to know the predicted execution time change that implies.

**Reply by makubacki** [reply]
I added the execution times for this change for reference since it triggered builds in all the packages: https://github.com/tianocore/edk2/pull/6343

**Reply by kraxel** [reply]
For OVMF / ArmVirt the most important build variant is DEBUG.
Not sure about NOOPT.  Sometimes catches build problems which slip through with optimizing builds, so I'd tend to keep them.
The RELEASE builds are not very important and can move to nightly pipeline IMHO.

I'd suggest to split the ovmf build pipeline like this:
* ovmf core (every PR):
  * OVMF IA32
  * OVMF X64
  * OVMF IA32X64
  * OVMF IA32X64 FULL
* ovmf platforms (ideally run on PRs which change OvmfPkg)
  * MICROVM
  * BHYVE
  * CLOUDHV
  * XEN
* ovmf confidential computing (run on OvmfPkg and CC changes).
  * AMDSEV
  * INTELTDX
* ovmf riscv (run on OvmfPkg and riscv platform changes)
  * RISCV64
* ovmf loongarch (run on OvmfPkg and loongarch platform changes)
  * LOONGARCH64

For ArmVirtPkg I'd say the `QEMU AARCH64` and `QEMU ARM` builds are core, the other ones can be moved to an 'armvirt platform' pipeline which runs only in case ArmVirtPkg code is changed.  Maybe also in case ArmPkg or ArmPlatformPkg is changed.

**Reply by os-d** [reply]
@kraxel , thanks for the breakdown, this is generally along the lines I’ve been thinking. One question I have is whether IA32 and ARM32 should be on every PR, new platforms that are exclusively IA32/ARM are pretty few and far between, right?

**Reply by mdkinney** [reply]
The IA32 coverage is important because of the use of intrinsics when 64-bit math is performed.  Those work without issue on X64 builds.

**Reply by makubacki** [reply]
This looks like a good list to me too.

I didn't mention that the `NO-TARGET` build runs in the [Windows VS2019 PR](https://dev.azure.com/tianocore/edk2-ci/_build?definitionId=34&_a=summary) (~7 min) and [Ubuntu GCC5](https://dev.azure.com/tianocore/edk2-ci/_build?definitionId=33&_a=summary) (~8 min) pipelines (based on need) today. That would continue to be the case.

**Reply by os-d** [reply]
> The IA32 coverage is important because of the use of intrinsics when 64-bit math is performed. Those work without issue on X64 builds.

Sure, for IA32 SEC/PEI, that would be covered in the IA32X64 Ovmf run. No one has IA32 DXE/SMM anymore, right? Or are there still new exclusively IA32 platforms?

**Reply by mdkinney** [reply]
It would be for coverage for common libs in case they are used in PEI by other platforms or might be used by PEI in the future.

However, if there is IA32 coverage for package builds, then I agree the OVMF IA32 build could be moved to once a day

**Reply by kraxel** [reply]
> @kraxel , thanks for the breakdown, this is generally along the lines I’ve been thinking. One question I have is whether IA32 and ARM32 should be on every PR, new platforms that are exclusively IA32/ARM are pretty few and far between, right?

I've only splitted the existing tests into groups ...

Going to figure what exactly makes sense to test going forward (and the related topic of the future of 32-bit platforms in edk2 in general) is a discussion we should have too, but this is IMHO much more than a CI tweak.

But as you are asking here is what I think:
* For OVMF it makes sense to shift the focus from OvmfIa32X64 to OvmfX64.  Roughly one year ago the pure 64-bit build gained feature parity with the 32-bit PEI / 64-bit DXE/SMM build.  There is no reason any more to prefer the mixed build over the pure 64-bit build.
* As already mentioned OVMF also has the role of a test vehicle for the edk2 code base.  So as long as edk2 wants continue supporting 32-bit platforms (either the pure 32-bit builds, or the mixed 32-bit PEI / 64-DXE builds, or both) we should continue to run those builds in CI.  And I also think it makes sense to run the tests on every PR because 32-bit only build failures are a relatively common problem (intrinsics, getting casts wrong, ...).  Allowing such problems creep into the code base is a bad idea I think.
